### ISSUE TITLE
[RFC]: Add DSpark speculative decoding support for DeepSeek-V4

### ISSUE THREAD

**Issue by EanWang211123** [vllm-project/vllm-ascend#11126]
### Motivation.
DeepSeek released DSpark speculative decoding for DeepSeek-V4, with official checkpoints:
- https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-DSpark
- https://huggingface.co/deepseek-ai/DeepSeek-V4-Pro-DSpark
DeepSeek reports **60–85%** faster per-user generation vs MTP-1 at matched throughput (Flash).
vllm-ascend already supports V4 + MTP, but DSpark is not yet a stable spec-decode method. Draft work exists in [#11066](https://github.com/vllm-project/vllm-ascend/pull/11066).
References:
- https://github.com/deepseek-ai/DeepSpec
- Upstream: https://github.com/vllm-project/vllm/issues/46910

### Proposed Change.

Add DSpark support for DeepSeek-V4 on Ascend:
1. Load official `*-DSpark` checkpoints (Markov head + confidence head).
2. Implement `AscendDsparkProposer` (propose → dynamic verify → reject sample).
3. Expose via `--speculative-config` (prefer `method: "dspark"`, or extend existing MTP path).

**Comment by QwertyJack** [comment]
## Design update: DeepSeek V4 DSpark architecture and four-PR delivery

This comment replaces the previous seven-PR plan. The end-to-end prototype in #11196 validated the main DeepSeek V4 DSpark functional and performance path, while #11765 has since added the generic Qwen/GLM DSpark proposer and shared input-preparation infrastructure to `main`.

The DeepSeek V4 implementation will therefore extend the generic DSpark foundation already present on the latest `main`, rather than introducing a second near-duplicate proposer. The remaining work is organized as four reviewable PRs covering the attention contract, model contract, eager correctness path, and graph optimization path.

### Goals

The implementation will provide:

- DSpark draft-model loading from the target checkpoint's `mtp.*` namespace.
- Target auxiliary hidden-state extraction from selected DeepSeek V4 layers.
- Anchor-first, noncausal block drafting followed by sequential Markov sampling.
- Greedy and probabilistic draft sampling with logits-space rejection recovery.
- Standard DeepSeek V4 DSA/SWA paged KV-cache integration.
- `SparseAttnSharedkv` SAS mode 4 with explicit `ori_sparse_indices`.
- BF16 and Ascend W4A8 model paths, including ModelSlim MTP mapping, optional QuaRot metadata, and DSpark FP8 QDQ semantics.
- DSpark drafter FULL ACLGraph capture/replay when the target model supports FULL graphs.

### Design invariants

1. **Standard paged cache is authoritative.**
   DSpark query and context KV use the standard DSA/SWA paged cache. The implementation does not introduce a second request-local KV-cache contract.

2. **The current draft block is noncausal.**
   Each query token attends to the historical sliding window and the complete current draft block. Explicit visible slots are represented by `ori_sparse_indices`; the band window is only the scheduling bound.

3. **Sampling semantics are independent of the attention backend.**
   The proposer produces base draft logits, applies the Markov transition sequentially, and preserves the processed draft distribution required by probabilistic rejection sampling.

4. **Correctness is established before graph optimization.**
   The eager proposer is the correctness baseline. FULL ACLGraph capture/replay is added only after the eager path is independently validated.

5. **Existing paths remain unchanged when DeepSeek V4 DSpark is disabled.**
   Generic Qwen/GLM DSpark from #11765, DeepSeek V4 without DSpark, existing MTP/EAGLE/DFlash methods, causal DSA/SWA, and existing `SparseAttnSharedkv` calls must preserve their behavior.

6. **Invalid metadata fails closed.**
   Missing standard-cache metadata, inconsistent request/block mappings, invalid speculative ranges, and incomplete request rows raise explicit errors rather than silently changing cache or position semantics.

### End-to-end data flow

1. The target DeepSeek V4 model captures auxiliary hidden states from configured layers.
2. The DeepSeek V4 specialization of the generic DSpark proposer converts scheduled target tokens into one fixed-size anchor-first draft block per request.
3. Target hidden states precompute and store DSpark context KV in the standard SWA paged cache.
4. The draft model computes block queries and current-block KV.
5. DSA metadata builders construct per-request query ranges, per-group block tables and slot mappings, and explicit SWA indices.
6. `SparseAttnSharedkv` reads the historical window and complete current block using `ori_sparse_indices`.
7. The Markov head samples the draft block sequentially and retains processed draft logits when probabilistic sampling is active.
8. Target verification uses the normal speculative-decoding path and logits-space rejection sampling.
9. In FULL ACLGraph mode, the fixed-shape draft forward is captured and replayed with persistent buffers; dynamic input preparation and final output slicing remain outside the graph.

### Four-PR delivery plan

The estimates include owned tests and are relative to the latest `main` after reusing #11765.

| PR | Proposed title | Main responsibility | Estimated size | Dependency |
| --- | --- | --- | ---: | --- |
| 1 | `[Feature][Attention] Add DeepSeek V4 DSpark noncausal DSA/SAS support` | `ori_sparse_indices` kernel support plus noncausal DSA/SWA metadata | about 1,000 lines | None |
| 2 | `[Feature][Model] Add DeepSeek V4 DSpark draft model and loading hooks` | Target hidden states, W4A8 mapping, draft model, SAS adapter, and weight loading | about 1,100-1,300 lines | PR 1 |
| 3 | `[Feature][SpecDecode] Add DeepSeek V4 DSpark eager decoding and probabilistic verification` | DeepSeek V4 specialization of the generic proposer, eager orchestration, Markov sampling, and logits-space rejection | about 1,000-1,400 lines | PR 2 |
| 4 | `[Performance][SpecDecode] Add FULL ACLGraph for DeepSeek V4 DSpark drafter` | Stable graph buffers, capture/replay, DP row padding, graph-mode synchronization, and replay slicing | about 900-1,200 lines | PR 3 |

Dependency order:

```text
PR 1: attention contract
  -> PR 2: model and loading contract
  -> PR 3: eager speculative-decoding correctness
  -> PR 4: FULL ACLGraph optimization
```

### PR 1: noncausal DSA/SAS attention contract

Implementation:

- Accept and validate original-KV sparse indices in `SparseAttnSharedkv`.
- Propagate sparse-index shape, dtype, and width through tiling.
- Use explicit physical KV slots in the SWA kernel path.
- Build the DeepSeek V4 DSpark noncausal visibility metadata for DSA and DSA context parallel.
- Keep causal and non-DSpark paths unchanged.

Validation:

- Existing calls without `ori_sparse_indices` preserve their output.
- Explicit-slot output matches a reference gather plus attention implementation.
- DSA and DSA context-parallel metadata agree on golden visibility cases.
- Cover non-contiguous slots, cross-block slots, padding, zero-length requests, and invalid shapes.
- Rebuild the final custom operator and run the single-card NPU parity matrix.

### PR 2: DeepSeek V4 model and loading contract

Implementation:

- Extract configured auxiliary hidden states from the target model.
- Add auditable DeepSeek V4 decoder, MoE, and draft-weight mapping hooks.
- Map ModelSlim target-layer prefixes to the checkpoint's `model.mtp.*` namespace.
- Add the Ascend DeepSeek V4 DSpark draft model and standard-cache SAS adapter.
- Load BF16/W4A8 weights, optional QuaRot metadata, FP8 QDQ parameters, embeddings, lm-head, and Markov-head weights.

Validation:

- Non-DSpark DeepSeek V4 behavior remains unchanged.
- Hidden-state shape, layer order, and padding are correct.
- Representative ModelSlim mappings and unmatched-key failures are covered.
- The real W4A8 checkpoint loads completely and one real model forward returns valid hidden states and logits.

### PR 3: eager DeepSeek V4 DSpark and probabilistic verification

Implementation:

- Reuse the generic `AscendDsparkProposer` and shared infrastructure from #11765.
- Add only the DeepSeek V4-specific standard DSA-cache, per-group metadata, and target/draft handoff behavior.
- Build anchor-first draft inputs and request-local query ranges.
- Run sequential Markov sampling for greedy and probabilistic requests.
- Preserve request-aligned processed draft logits for logits-space rejection sampling.
- Harden generic rejection and padded-input paths required by the validated DeepSeek V4 flow.

Validation:

- Fixed-seed logits-space parity against a probability-space rejection reference.
- Existing Qwen/GLM DSpark, EAGLE, MTP, DFlash, n-gram, 310P, and v2 paths do not regress.
- Real W4A8 eager service returns `/v1/models` and a non-empty chat completion.
- Cover rejected anchors, maximum model length, request reordering, mixed greedy/random batches, multi-request padding, and uneven DP turnover.
- A fixed coding workload reports healthy AR, AL, and per-position acceptance.

### PR 4: DeepSeek V4 DSpark FULL ACLGraph

Implementation:

- Derive DeepSeek V4 DSpark draft capture sizes from target request rows.
- Use persistent graph input, attention metadata, sampling, and output buffers.
- Pad complete inactive request rows across DP ranks.
- Synchronize only graph modes that are safe under uneven DP request turnover.
- Replay the fixed-shape draft forward and slice outputs back to actual requests.
- Keep `speculative_config.enforce_eager=true` as the eager isolation/fallback path.

Validation:

- Eager-versus-ACLGraph deterministic token/logit parity on controlled probes.
- Greedy and probabilistic-path parity.
- Explicit capture, replay, SAS execution, and stable-address evidence.
- Real-weight TP4/DP4/EP16 validation on 16 NPUs.
- Full SPEED-Bench `671/671` completion with zero failed requests.
- Repeated coding80 runs with AR, AL, and per-position acceptance.
- Fixed 8k input / 1k output performance at BS 1, 4, 8, 16, and 32.

### End-to-end acceptance criteria

The final cumulative branch will be validated using the same real W4A8 checkpoint and topology used for #11196:

- 16 NPUs with TP4 / DP4 / EP16.
- Standard paged SWA cache, standard DSA, and SAS mode 4.
- Five speculative tokens, with the drafter FULL ACLGraph enabled by default.
- `/v1/models` and a real `/v1/chat/completions` request return HTTP 200 with non-empty output.
- Full SPEED-Bench completes `671/671` requests with zero failures.
- Coding80 reports AR, AL, and per-position acceptance aligned with #11196 within normal run-to-run variance.
- The fixed 8k/1k BS 1/4/8/16/32 matrix demonstrates the same end-to-end performance benefit as #11196 under matched launch and benchmark settings.
- Logs preserve import-root evidence, KV-cache capacity, ACLGraph capture/replay, SAS execution, request failures, and post-run cleanup state.

Dummy weights and reduced-layer checkpoints remain engineering probes only; they do not replace real-weight correctness, acceptance, or performance sign-off.

### Non-goals

- RNN-head support.
- Confidence-head calibration or dynamic verification length.
- Hardware-aware prefix scheduling.
- New DeepSeek V4 DSpark-specific user-facing environment variables.
- Unrelated KV-write, SAS tiling, or communication optimizations.
- Offline BF16 checkpoint conversion tooling.

### References

- DeepSeek V4 DSpark RFC: this issue.
- Broad DSpark roadmap: #11163.
- Generic Qwen/GLM DSpark foundation: #11765.
- End-to-end DeepSeek V4 Ascend prototype: #11196.
- Upstream vLLM DSpark implementation: https://github.com/vllm-project/vllm/pull/46995.

Each implementation PR will reference this RFC, its direct dependency, the latest `main` base, and the validation evidence owned by that PR.
